File: e_2_develop_idea/src/main/java/com/sj/e_2_develop_idea/MainActivity.java
package com.sj.e_2_develop_idea;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.sj.e_2_develop_idea.workflow.Work;
import com.sj.e_2_develop_idea.workflow.WorkFlow;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.hello_tv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startFlow();
            }
        });
    }

    /**
     * 工作流方式：方便任务间的调整，消除强耦合
     * 参考链接：https://mp.weixin.qq.com/s/sKRe_jyFBYEDIKdgjz7rZQ
     */
    private void startFlow() {
        WorkFlow workFlow = new WorkFlow(new WorkFlow.WorkFlowListener() {
            @Override
            public void flowEnd(String msg) {
                Toast.makeText(MainActivity.this,msg,Toast.LENGTH_SHORT).show();
            }
        });
        workFlow.addWork(new Work("1", new Work.Worker() {
            @Override
            public void workDuties(Work work) {
                Toast.makeText(MainActivity.this,"Task 1",Toast.LENGTH_SHORT).show();
                //具体任务--完成后调用完成
                work.workComplete();
            }
        }));
        workFlow.addWork(new Work("2", new Work.Worker() {
            @Override
            public void workDuties(Work work) {
                Toast.makeText(MainActivity.this,"Task 2",Toast.LENGTH_SHORT).show();
                //具体任务--完成后调用完成
                work.workComplete();
            }
        }));
        workFlow.addWork(new Work("4", new Work.Worker() {
            @Override
            public void workDuties(Work work) {
                Toast.makeText(MainActivity.this,"Task 4",Toast.LENGTH_SHORT).show();
                //具体任务--完成后调用完成
                work.workComplete();
            }
        }));
        workFlow.addWork(new Work("3", new Work.Worker() {
            @Override
            public void workDuties(Work work) {
                Toast.makeText(MainActivity.this,"Task 3",Toast.LENGTH_SHORT).show();
                //具体任务--完成后调用完成
                work.workComplete();
            }
        }));
        workFlow.startWork();
    }
}

File: a_2_service/src/main/java/com/sj/a_2_service/MainActivity.kt
package com.sj.a_2_service

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    var TAG = "MainActivity"
    var count=0;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        start_simple_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                var intent=Intent(this@MainActivity,SimpleService::class.java)
                startService(intent)
            }
        })
        stop_simple_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                var intent=Intent(this@MainActivity,SimpleService::class.java)
                intent.action="android.intent.action.SimpleService"
                intent.`package`=this@MainActivity.packageName
                stopService(intent)
            }
        })
        bind_simple_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                var intent=Intent(this@MainActivity,SimpleService::class.java)
                count++
                intent.type="c"+count
                bindService(intent,serviceConnection, Context.BIND_AUTO_CREATE)
            }
        })
        unbind_simple_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                unbindService(serviceConnection)
            }
        })
        start_foreground_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                startService(Intent(this@MainActivity,ForegroundService::class.java))
            }
        })
        start_intent_service_tv.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                startService(Intent(this@MainActivity,MyIntentService::class.java))
            }
        })
    }
    lateinit var binder: SimpleService.CountBinder
    var serviceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(p0: ComponentName?) {
            Log.e(TAG,"异常解绑时调用")
        }
        override fun onServiceConnected(p0: ComponentName?, p1: IBinder?) {
            Log.e(TAG,"成功绑定")
            binder = p1 as SimpleService.CountBinder;
            binder.startRecord()
            binder.stopRecord()
            Log.e(TAG,""+binder.count+"")
        }
    }
}
